File: packages/uns-database/src/params.ts
import type {
  CompiledSqlStatement,
  DatabaseDialect,
  SqlParams,
} from "./types.js";

export const ORACLE_IN_LIST_LIMIT = 1000;

const isIdentifierStart = (char: string | undefined): boolean =>
  !!char && /[A-Za-z_]/.test(char);

const isIdentifierPart = (char: string | undefined): boolean =>
  !!char && /[A-Za-z0-9_]/.test(char);

function readParameterName(sqlText: string, startIndex: number): { name: string; endIndex: number } {
  let index = startIndex;
  let name = "";

  while (index < sqlText.length && isIdentifierPart(sqlText[index])) {
    name += sqlText[index];
    index += 1;
  }

  return { name, endIndex: index - 1 };
}

export function compileNamedParams(
  dialect: DatabaseDialect,
  sqlText: string,
  params: SqlParams = {}
): CompiledSqlStatement {
  const output: string[] = [];
  const values: unknown[] = [];
  const bindValues: Record<string, unknown> = {};
  const parameterOrder: string[] = [];
  const pgIndexes = new Map<string, number>();

  let inSingleQuote = false;
  let inDoubleQuote = false;
  let inLineComment = false;
  let inBlockComment = false;

  for (let index = 0; index < sqlText.length; index += 1) {
    const char = sqlText[index];
    const next = sqlText[index + 1];
    const previous = sqlText[index - 1];

    if (inLineComment) {
      output.push(char);
      if (char === "\n") {
        inLineComment = false;
      }
      continue;
    }

    if (inBlockComment) {
      output.push(char);
      if (char === "*" && next === "/") {
        output.push(next);
        inBlockComment = false;
        index += 1;
      }
      continue;
    }

    if (!inDoubleQuote && char === "'" && previous !== "\\") {
      inSingleQuote = !inSingleQuote;
      output.push(char);
      continue;
    }

    if (!inSingleQuote && char === "\"" && previous !== "\\") {
      inDoubleQuote = !inDoubleQuote;
      output.push(char);
      continue;
    }

    if (inSingleQuote || inDoubleQuote) {
      output.push(char);
      continue;
    }

    if (char === "-" && next === "-") {
      output.push(char, next);
      inLineComment = true;
      index += 1;
      continue;
    }

    if (char === "/" && next === "*") {
      output.push(char, next);
      inBlockComment = true;
      index += 1;
      continue;
    }

    if (char === ":" && previous !== ":" && isIdentifierStart(next)) {
      const { name, endIndex } = readParameterName(sqlText, index + 1);

      if (!(name in params)) {
        throw new Error(`Missing SQL parameter ':${name}'.`);
      }

      const value = params[name];

      if (Array.isArray(value)) {
        const inListContext = dialect === "oracle"
          ? getInListContext(output, sqlText, endIndex + 1)
          : null;

        if (value.length === 0) {
          if (inListContext) {
            output.splice(inListContext.startIndex);
            output.push("1 = 0");
            index = inListContext.closingParenIndex;
          } else {
            output.push("(select 1 where 1 = 0)");
            index = endIndex;
          }
          continue;
        }

        const placeholders = buildArrayPlaceholders(
          dialect,
          name,
          value,
          values,
          bindValues,
          parameterOrder,
          inListContext?.leftSideExpression
        );

        if (inListContext) {
          output.splice(inListContext.startIndex);
        }

        output.push(placeholders);
        if (inListContext) {
          index = inListContext.closingParenIndex;
        } else {
          index = endIndex;
        }
        continue;
      }

      parameterOrder.push(name);

      if (dialect === "oracle") {
        bindValues[name] = value;
        output.push(`:${name}`);
      } else if (dialect === "pg") {
        let paramIndex = pgIndexes.get(name);
        if (!paramIndex) {
          values.push(value);
          paramIndex = values.length;
          pgIndexes.set(name, paramIndex);
        }
        output.push(`$${paramIndex}`);
      } else {
        values.push(value);
        output.push("?");
      }

      index = endIndex;
      continue;
    }

    output.push(char);
  }

  return {
    text: output.join(""),
    values: dialect === "oracle" ? bindValues : values,
    parameterOrder,
  };
}

function buildArrayPlaceholders(
  dialect: DatabaseDialect,
  name: string,
  arrayValue: unknown[],
  values: unknown[],
  bindValues: Record<string, unknown>,
  parameterOrder: string[],
  leftSideExpression?: string
): string {
  if (dialect === "oracle") {
    const chunks = chunkArray(arrayValue, ORACLE_IN_LIST_LIMIT);
    const inColumn = leftSideExpression?.trim();

    const chunkSql = chunks.map((chunk, chunkIndex) => {
      const placeholders = chunk.map((item, itemIndex) => {
        const bindName = `${name}_${chunkIndex}_${itemIndex}`;
        bindValues[bindName] = item;
        parameterOrder.push(bindName);
        return `:${bindName}`;
      });

      return inColumn
        ? `${inColumn} IN (${placeholders.join(", ")})`
        : placeholders.join(", ");
    });

    return inColumn
      ? `(${chunkSql.join(" OR ")})`
      : chunkSql.join(", ");
  }

  if (dialect === "pg") {
    const placeholders = arrayValue.map(item => {
      values.push(item);
      parameterOrder.push(name);
      return `$${values.length}`;
    });
    return placeholders.join(", ");
  }

  const placeholders = arrayValue.map(item => {
    values.push(item);
    parameterOrder.push(name);
    return "?";
  });
  return placeholders.join(", ");
}

function chunkArray<T>(values: T[], size: number): T[][] {
  const chunks: T[][] = [];

  for (let index = 0; index < values.length; index += size) {
    chunks.push(values.slice(index, index + size));
  }

  return chunks;
}

function getInListContext(
  output: string[],
  sqlText: string,
  searchStartIndex: number
): {
  startIndex: number;
  leftSideExpression: string;
  closingParenIndex: number;
} | null {
  const outputText = output.join("");
  const match = /([A-Za-z0-9_."$]+)\s+IN\s*\($/i.exec(outputText);

  if (!match) {
    return null;
  }

  let cursor = searchStartIndex;
  while (cursor < sqlText.length && /\s/.test(sqlText[cursor])) {
    cursor += 1;
  }

  if (sqlText[cursor] !== ")") {
    return null;
  }

  return {
    startIndex: match.index,
    leftSideExpression: match[1],
    closingParenIndex: cursor,
  };
}

File: packages/uns-py/src/uns_kit/database/params.py
from __future__ import annotations

from .types import CompiledSqlStatement, DatabaseDialect, SqlParams

ORACLE_IN_LIST_LIMIT = 1000


def _is_identifier_start(char: str | None) -> bool:
    return bool(char) and (char.isalpha() or char == "_")


def _is_identifier_part(char: str | None) -> bool:
    return bool(char) and (char.isalnum() or char == "_")


def _read_parameter_name(sql_text: str, start_index: int) -> tuple[str, int]:
    index = start_index
    name = []
    while index < len(sql_text) and _is_identifier_part(sql_text[index]):
        name.append(sql_text[index])
        index += 1
    return "".join(name), index - 1


def compile_named_params(
    dialect: DatabaseDialect,
    sql_text: str,
    params: SqlParams | None = None,
) -> CompiledSqlStatement:
    resolved_params = params or {}
    output: list[str] = []
    values: list[object] = []
    bind_values: dict[str, object] = {}
    parameter_order: list[str] = []
    in_single_quote = False
    in_double_quote = False
    in_line_comment = False
    in_block_comment = False

    index = 0
    while index < len(sql_text):
        char = sql_text[index]
        next_char = sql_text[index + 1] if index + 1 < len(sql_text) else None
        previous_char = sql_text[index - 1] if index > 0 else None

        if in_line_comment:
            output.append(char)
            if char == "\n":
                in_line_comment = False
            index += 1
            continue

        if in_block_comment:
            output.append(char)
            if char == "*" and next_char == "/":
                output.append("/")
                in_block_comment = False
                index += 2
                continue
            index += 1
            continue

        if not in_double_quote and char == "'" and previous_char != "\\":
            in_single_quote = not in_single_quote
            output.append(char)
            index += 1
            continue

        if not in_single_quote and char == '"' and previous_char != "\\":
            in_double_quote = not in_double_quote
            output.append(char)
            index += 1
            continue

        if in_single_quote or in_double_quote:
            output.append(char)
            index += 1
            continue

        if char == "-" and next_char == "-":
            output.extend(["-", "-"])
            in_line_comment = True
            index += 2
            continue

        if char == "/" and next_char == "*":
            output.extend(["/", "*"])
            in_block_comment = True
            index += 2
            continue

        if char == ":" and previous_char != ":" and _is_identifier_start(next_char):
            name, end_index = _read_parameter_name(sql_text, index + 1)
            if name not in resolved_params:
                raise ValueError(f"Missing SQL parameter ':{name}'.")
            value = resolved_params[name]
            if isinstance(value, list):
                in_list_context = _get_in_list_context("".join(output), sql_text, end_index + 1) if dialect == "oracle" else None
                if len(value) == 0:
                    if in_list_context is not None:
                        del output[in_list_context["start_index"] :]
                        output.append("1 = 0")
                        index = in_list_context["closing_paren_index"] + 1
                        continue
                    output.append("(select 1 where 1 = 0)")
                    index = end_index + 1
                    continue

                placeholders = _build_array_placeholders(
                    dialect,
                    name,
                    value,
                    values,
                    bind_values,
                    parameter_order,
                    in_list_context["left_side_expression"] if in_list_context else None,
                )
                if in_list_context is not None:
                    del output[in_list_context["start_index"] :]
                output.append(placeholders)
                index = (in_list_context["closing_paren_index"] + 1) if in_list_context else (end_index + 1)
                continue

            if dialect == "oracle":
                parameter_order.append(name)
                bind_values[name] = value
                output.append(f":{name}")
            elif dialect == "pg":
                values.append(value)
                parameter_order.append(name)
                output.append("%s")
            else:
                parameter_order.append(name)
                values.append(value)
                output.append("?")
            index = end_index + 1
            continue

        output.append(char)
        index += 1

    return CompiledSqlStatement(
        text="".join(output),
        values=bind_values if dialect == "oracle" else values,
        parameter_order=parameter_order,
    )


def _build_array_placeholders(
    dialect: DatabaseDialect,
    name: str,
    array_value: list[object],
    values: list[object],
    bind_values: dict[str, object],
    parameter_order: list[str],
    left_side_expression: str | None = None,
) -> str:
    if dialect == "oracle":
        chunks = [array_value[index : index + ORACLE_IN_LIST_LIMIT] for index in range(0, len(array_value), ORACLE_IN_LIST_LIMIT)]
        chunk_sql: list[str] = []
        for chunk_index, chunk in enumerate(chunks):
            placeholders = []
            for item_index, item in enumerate(chunk):
                bind_name = f"{name}_{chunk_index}_{item_index}"
                bind_values[bind_name] = item
                parameter_order.append(bind_name)
                placeholders.append(f":{bind_name}")
            if left_side_expression:
                chunk_sql.append(f"{left_side_expression.strip()} IN ({', '.join(placeholders)})")
            else:
                chunk_sql.append(", ".join(placeholders))
        return f"({' OR '.join(chunk_sql)})" if left_side_expression else ", ".join(chunk_sql)

    if dialect == "pg":
        placeholders = []
        for item in array_value:
            values.append(item)
            parameter_order.append(name)
            placeholders.append("%s")
        return ", ".join(placeholders)

    placeholders = []
    for item in array_value:
        values.append(item)
        parameter_order.append(name)
        placeholders.append("?")
    return ", ".join(placeholders)


def _get_in_list_context(output_text: str, sql_text: str, search_start_index: int) -> dict[str, int | str] | None:
    import re

    match = re.search(r'([A-Za-z0-9_."$]+)\s+IN\s*\($', output_text, re.IGNORECASE)
    if match is None:
        return None

    cursor = search_start_index
    while cursor < len(sql_text) and sql_text[cursor].isspace():
        cursor += 1
    if cursor >= len(sql_text) or sql_text[cursor] != ")":
        return None

    return {
        "start_index": match.start(),
        "left_side_expression": match.group(1),
        "closing_paren_index": cursor,
    }
